pkg/commands/copy_test.go

```diff
@@ -461,6 +462,60 @@ func TestCopyCommand_ExecuteCommand_Extended(t *testing.T) {
 		}
 	})
 
+	t.Run("copy dir to another dir - with ignored files", func(t *testing.T) {
+		testDir, srcDir := setupDirs(t)
+		defer os.RemoveAll(testDir)
+		ignoredFile := "bam.txt"
+		srcFiles, err := ioutil.ReadDir(filepath.Join(testDir, srcDir))
+		if err != nil {
+			t.Fatal(err)
+		}
+		expected := map[string]fs.FileInfo{}
+		for _, sf := range srcFiles {
+			if sf.Name() == ignoredFile {
+				continue
+			}
+			expected[sf.Name()] = sf
+		}
+
+		cmd := CopyCommand{
+			cmd: &instructions.CopyCommand{
+				SourcesAndDest: instructions.SourcesAndDest{SourcePaths: []string{srcDir}, DestPath: "dest"},
+			},
+			fileContext: util.FileContext{
+				Root:          testDir,
+				ExcludedFiles: []string{filepath.Join(srcDir, "bam.txt")}},
+		}
+
+		cfg := &v1.Config{
+			Cmd:        nil,
+			Env:        []string{},
+			WorkingDir: testDir,
+		}
+
+		err = cmd.ExecuteCommand(cfg, dockerfile.NewBuildArgs([]string{}))
+		if err != nil {
```

Comment:
Could you help me to understand what's the error as described in #1598 could have happened without this fix?

Reply:
Hi @JeromeJu. IIUC #1598 describes a racing condition between 2 process. Say process A 1) finds all the files to be copied; and 2) copies the files one by one. If process B kicks in and deletes some file between 1) and 2), it will throw an error. However, if the files to be copied are supposed to be ignored, it should not fail since the files will not be copied anyways. That's all swapping the 2 function calls solve the problem.

Regarding the tests, I didn't find a way to mock the race condition in unit test, so I just added a test case with ignored files. Please let me know if you have better way to test it 😁

Reply:
One way of doing this is via "monkey patching" (done elsewhere in the code) where we can mock out `RemainingFiles` or `os.Lstat` to immediately delete an excluded file after processing and use this mock for a repro test.  I have done this here - https://gist.github.com/aaron-prindle/7f7fd1f43dc5c3fe4e9b14f69ecdd9e7

I don't think it is necessary for the code in this case though, I have manually tested with the above and things seem to be working.  Looks good!